File: framework/languages/dotnet/samples/Bingo/Client/BingoClientApp.cs
using Bingo.Shared.Configuration;
using Bingo.Shared.Contracts;
using Systems.Zlink.Stream.Connector.Contracts;
using Systems.Zlink.Stream.Connector.Json;

namespace Bingo.Client;

internal sealed class BingoClientApp
{
    public async ValueTask RunAsync(
        IZlinkStreamConnector client1,
        IZlinkStreamConnector client2,
        CancellationToken cancellationToken = default)
    {
        // Client 1 connects, authenticates, and creates the waiting room.
        await client1.ConnectAsync(cancellationToken);
        var client1Auth = await client1
            .Request(new AuthenticateReq("player-1"))
            .Timeout(SampleTimings.RequestTimeout)
            .SubmitAsync<AuthenticateRes>(cancellationToken);

        Require(client1Auth.ActorId == "player-1", "Client 1 must authenticate as player-1.");

        var client1MatchRes = await client1
            .Request(new MatchBingoReq("two-player"))
            .Timeout(SampleTimings.RequestTimeout)
            .SubmitAsync<MatchBingoRes>(cancellationToken);

        Require(client1MatchRes.State.Status == "WaitingForPlayers", "Client 1 match must leave room waiting.");
        Require(client1MatchRes.State.HostActorId == client1Auth.ActorId, "Client 1 must become room host.");
        Require(client1.ReceivedCount(nameof(PlayerJoinedNotify)) == 0, "Client 1 must not receive a self-join notification.");

        // Client 2 connects, authenticates, and joins the same room.
        await client2.ConnectAsync(cancellationToken);

        var client2Auth = await client2
            .Request(new AuthenticateReq("player-2"))
            .Timeout(SampleTimings.RequestTimeout)
            .SubmitAsync<AuthenticateRes>(cancellationToken);

        Require(client2Auth.ActorId == "player-2", "Client 2 must authenticate as player-2.");
        Require(client2Auth.ActorId != client1Auth.ActorId, "Clients must authenticate as distinct actors.");

        var client2MatchRes = await client2
            .Request(new MatchBingoReq("two-player"))
            .Timeout(SampleTimings.RequestTimeout)
            .SubmitAsync<MatchBingoRes>(cancellationToken);

        Require(client2MatchRes.RoomId == client1MatchRes.RoomId, "Both clients must match into one room.");
        Require(client2MatchRes.State.Status == "Running", "Client 2 match must start the room.");

        // Joining another player is delivered as a push to existing room members.
        var client1SawClient2Join = await client1.WaitForAsync<PlayerJoinedNotify>(
            SampleTimings.RequestTimeout,
            cancellationToken);

        Require(client1SawClient2Join.Payload.ActorId == client2Auth.ActorId, "Client 1 must be notified when client 2 joins.");
        Require(client2.ReceivedCount(nameof(PlayerJoinedNotify)) == 0, "Client 2 must not receive a self-join notification.");

        // The room starts automatically after both players have joined.
        var client1StartedTask = client1.WaitForAsync<BingoGameStartedNotify>(
            SampleTimings.RequestTimeout,
            cancellationToken).AsTask();
        var client2StartedTask = client2.WaitForAsync<BingoGameStartedNotify>(
            SampleTimings.RequestTimeout,
            cancellationToken).AsTask();
        await Task.WhenAll(client1StartedTask, client2StartedTask);

        var client1Started = await client1StartedTask;
        Require(client1Started.Payload.State.Status == "Running", "Client 1 must receive game start.");

        var client2Started = await client2StartedTask;
        Require(client2Started.Payload.State.Status == "Running", "Client 2 must receive game start.");

        // Each client submits its bingo card; the server starts drawing after both cards arrive.
        var client2Card = await client2
            .Request(new SubmitBingoCardReq(client2MatchRes.RoomId, BingoClientCards.Player2))
            .Timeout(SampleTimings.RequestTimeout)
            .SubmitAsync<SubmitBingoCardRes>(cancellationToken);
        Require(client2Card.State.Status == "Running", "Client 2 card submission must keep room running.");

        var client1Card = await client1
            .Request(new SubmitBingoCardReq(client1MatchRes.RoomId, BingoClientCards.Player1))
            .Timeout(SampleTimings.RequestTimeout)
            .SubmitAsync<SubmitBingoCardRes>(cancellationToken);
        Require(client1Card.State.Status == "Running", "Client 1 card submission must keep room running.");

        var drawnNumbers = new List<BingoNumberDrawnNotify>();
        // Number drawing is server-driven; clients only wait for draw notifications.
        while (true)
        {
            cancellationToken.ThrowIfCancellationRequested();
            var expectedSeq = drawnNumbers.Count + 1;
            var client1DrawnTask = client1.WaitForAsync<BingoNumberDrawnNotify>(
                message => message.Payload.DrawSeq == expectedSeq,
                SampleTimings.RequestTimeout,
                cancellationToken).AsTask();
            var client2DrawnTask = client2.WaitForAsync<BingoNumberDrawnNotify>(
                message => message.Payload.DrawSeq == expectedSeq,
                SampleTimings.RequestTimeout,
                cancellationToken).AsTask();
            await Task.WhenAll(client1DrawnTask, client2DrawnTask);

            var client1Drawn = await client1DrawnTask;
            var client2Drawn = await client2DrawnTask;
            drawnNumbers.Add(client1Drawn.Payload);
            Require(client2Drawn.Payload.DrawSeq == client1Drawn.Payload.DrawSeq, "Both clients must receive the same draw sequence.");
            Require(client2Drawn.Payload.Number == client1Drawn.Payload.Number, "Both clients must receive the same drawn number.");

            if (client1Drawn.Payload.State.Status == "Finished")
            {
                break;
            }
        }
        Require(drawnNumbers.Count > 0, "Server timer must draw numbers.");
        Require(drawnNumbers[^1].State.Status == "Finished", "Server timer must finish the room.");

        // The final result is pushed to both clients when the server detects bingo.
        var client1ResultTask = client1.WaitForAsync<BingoGameEndedNotify>(
            message => message.Payload.State.Status == "Finished",
            SampleTimings.RequestTimeout,
            cancellationToken).AsTask();
        var client2ResultTask = client2.WaitForAsync<BingoGameEndedNotify>(
            message => message.Payload.State.Status == "Finished",
            SampleTimings.RequestTimeout,
            cancellationToken).AsTask();
        await Task.WhenAll(client1ResultTask, client2ResultTask);

        var client1Result = await client1ResultTask;
        Require(client1Result.Payload.State.Status == "Finished", "Client 1 must receive game result.");
        var client2Result = await client2ResultTask;
        Require(client2Result.Payload.State.Status == "Finished", "Client 2 must receive game result.");

        Require(
            client1Result.Payload.State.DrawnNumbers.SequenceEqual(drawnNumbers.Select(static draw => draw.Number)),
            "Final state must contain the drawn number sequence.");
        Require(
            client1Result.Payload.State.Winners.SequenceEqual([client1Auth.ActorId]),
            "The deterministic sample must finish with client 1 as the winner.");
        Require(
            client1Result.Payload.State.Players.All(static player => player.Card.Length == 9),
            "Each player card must contain 9 cells.");
        Require(
            client1Result.Payload.State.Players.All(static player => player.Marks[4]),
            "Center free cell must start marked.");
    }

    private static void Require(bool condition, string message)
    {
        if (!condition)
        {
            throw new InvalidOperationException(message);
        }
    }
}

internal static class BingoClientCards
{
    public static readonly int[] Player1 = [1, 2, 3, 4, 0, 6, 7, 8, 9];

    public static readonly int[] Player2 = [10, 11, 12, 13, 0, 14, 4, 5, 6];
}

File: framework/languages/node/samples/Bingo/Client/bingo-client-app.js
const { createChannelClient } = require('../../shared/channel-runtime');
const { BingoPlayerClient } = require('./bingo-player-client');
const { SampleNames, SampleTimings } = require('../Shared/Configuration/sample-names');

class BingoClientApp {
  async run(options) {
    const apiClient = await createChannelClient({
      channelName: SampleNames.apiChannel,
      peers: [options.apiEndpoint]
    });
    const playClient = await createChannelClient({
      channelName: SampleNames.playChannel,
      peers: [options.playEndpoint]
    });

    try {
      const clients = SampleNames.actorIds.map((actorId) => new BingoPlayerClient(actorId, apiClient, playClient));
      const authentications = [];
      for (const client of clients) {
        authentications.push(await client.authenticate());
      }

      const firstMatch = await clients[0].match();
      const earlyHostStartRejected = await isRejected(() => clients[0].start(firstMatch.roomId));
      const matches = [firstMatch];
      for (const client of clients.slice(1)) {
        matches.push(await client.match());
      }

      const nonHostStartRejected = await isRejected(() => clients[1].start(firstMatch.roomId));
      const started = await clients[0].start(firstMatch.roomId);
      const timer = await playClient.request('RunBingoRoomTimerReq', {
        roomId: firstMatch.roomId
      }, SampleTimings.requestTimeout);
      const delivered = (await playClient.request('BingoDeliveredNotificationsReq', {}, SampleTimings.requestTimeout)).delivered;
      for (const client of clients) {
        client.notifications.apply(delivered);
      }

      const result = {
        authentications,
        matches,
        started,
        ended: timer.state,
        playerJoinedPushCounts: clients.map((client) => client.notifications.playerJoined.length),
        startedPushCounts: clients.map((client) => client.notifications.started.length),
        drawnPushCounts: clients.map((client) => client.notifications.drawn.length),
        endedPushCounts: clients.map((client) => client.notifications.ended.length),
        earlyHostStartRejected,
        nonHostStartRejected
      };
      validate(result);
      return result;
    } finally {
      await playClient.stop();
      await apiClient.stop();
    }
  }
}

async function isRejected(action) {
  try {
    await action();
    return false;
  } catch {
    return true;
  }
}

function validate(result) {
  requireCondition(new Set(result.authentications.map((auth) => auth.actorId)).size === 4, 'Four clients must authenticate as distinct actors.');
  requireCondition(new Set(result.matches.map((match) => match.roomId)).size === 1, 'All match requests must return the same room.');
  requireCondition(result.matches[0].state.hostActorId === result.authentications[0].actorId, 'First joined actor must become host.');
  requireCondition(result.earlyHostStartRejected, 'Host start must be rejected before four players join.');
  requireCondition(result.nonHostStartRejected, 'Non-host start must be rejected.');
  requireCondition(result.started.state.status === 'Running', 'Host start must put room into Running status.');
  requireCondition(result.ended.status === 'Finished', 'Room must finish through timer draws.');
  requireCondition(result.ended.winners.length > 1, 'The deterministic sample must include same-sequence winners.');
  requireCondition(result.ended.players.every((player) => player.card.length === 25), 'Each player card must contain 25 cells.');
  requireCondition(result.ended.players.every((player) => player.marks[12]), 'Center free cell must start marked.');
  requireCondition(result.startedPushCounts.every((count) => count > 0), 'Each client must receive game-start push.');
  requireCondition(result.drawnPushCounts.every((count) => count > 0), 'Each client must receive draw push.');
  requireCondition(result.endedPushCounts.every((count) => count > 0), 'Each client must receive game-ended push.');
}

function requireCondition(condition, message) {
  if (!condition) {
    throw new Error(message);
  }
}

module.exports = { BingoClientApp };
